### PR TITLE
Use case: satellite / terrestrial multipath communication

### PR DESCRIPTION
**Pull request by joergdeutschmann-i7**
Similar to Hybrid Access, we have a use-case which combines low latency, low data rate and high data rate, high latency (geostationary satellite) links.
We need one or two minutes for presentation at most. The mentioned draft will be available within the next days.

### CODE REVIEW

**Review by joergdeutschmann-i7 — COMMENTED** [review]
Sorry, should go into interim-20-10 folder, of course...

**Comment by larseggert** [comment]
Please update your PR to add the PDF to the correct folder.

Also, I haven't discussed this with @LPardue yet, but here are my suggestions:
* On your first slide, where to you envision mutipath QUIC to be deployed? I guess on the proxy?
* I think you can skip slide 2
* I'm missing any concrete requirements you have for this use case - what is missing from QUICv1 to support it?

**Comment by joergdeutschmann-i7** [comment]
Thanks Lars for your feedback!
 

>     * On your first slide, where to you envision mutipath QUIC to be deployed? I guess on the proxy?
We actually think of using MPQUIC end-to-end and enabling this use case without the need for proxies. However, we don't have a MPQUIC implementation yet.


>     * I think you can skip slide 2
We can skip slide 2 during the meeting. If possible, I'd still like to keep it in the slide deck to show interested readers the possible benefits (maybe move this slide to the end / appendix).

 
>     * I'm missing any concrete requirements you have for this use case - what is missing from QUICv1 to support it?
The difficult part is (like so often) the scheduling. We frequently switch data from flows among both links (connection setup -> terrestrial, large chunks of data -> satellite, small data (interactive content?) -> terrestrial, etc.). I'm not sure whether this is efficiently realizable with QUICv1 connection migration. It's probably easier if both links are available at the same time.

**Comment by larseggert** [comment]
On the first point, then please change the slide then? Because at the moment, the "multipath protocol" is shown as running on the PEP, which is very confusing. Or add a slide that shows the desired future architecture.

If you intend to run multipath QUIC end-to-end, would you even need "your" own implementation? (I assume you are saying "we" the PEP vendor here.)

Yes, moving slide 2 to an appendix is OK.

You should explain the details of this uncertainty of whether v1 connection migration is sufficient for this use case in the slide deck. I think that is exactly the kind of information we would want to see as input to this discussion.

**Comment by LPardue** [comment]
Agree with Lars on all points. If you have a target architecture please present that; you can keep your test architecure alongside the test results if you think it helps that angle.

**Comment by joergdeutschmann-i7** [comment]
In the current (TCP) architecture, multipath is only present between the PEPs. However, with QUIC, PEPs are a thing of the past. That's our motivation for MPQUIC. I've tried to clarify this in the updated slides and tried to include your suggestions. Again, your feedback is highly appreciated.

**Comment by LPardue** [comment]
Thanks, that's a bit clearer. 

MPQUIC is a specific proposal for multipath QUIC. It would help if you can clarify on slide 2 if you'd like an abstract multipath capability or if you specifically need MPQUIC and uniflows.

**Comment by joergdeutschmann-i7** [comment]
Hmmm, not sure what you mean by abstract multipath capability. Because the congestion and flow control for both links is very different, I think that a multipath protocol needs to handle both links individually.

Maybe another way of thinking of it is comparing it to MPTCP. A scheduler for such a link combination could also be implemented in MPTCP, i.e. a scheduler which is aware of both (very different) subflows.
(Unfortunately we do not have our use case implemented in MPTCP. A implementation based on PEPs seemed easier, as these are deployed in satellite networks anyway...)

**Comment by LPardue** [comment]
What I mean is, if you replace "MPQUIC" with "multipath QUIC" and "uniflow" with "path" does your digram still communicate your intent?

What do you require from the transport to realize your goals? Would it be enough to write a scheduler that simply operates at the server application layer? (I suspect the answer is no, but the slides don't tell me if bandwidth aggregation is enough or if you need something in the protocol specifically targetted at paths with very different properties).

**Comment by joergdeutschmann-i7** [comment]
So you mean two QUIC paths managed by an server application layer? Well, I think the transport layer is the only sensible entity which has knowledge of the very different link characteristics.
Regarding bandwidth aggregation; this might be important as soon as the terrestrial link has at least some noteworthy capacity, so that both terrestrial and satellite link are used simultaneously.
If we assume a terrestrial link with a really low data rate, connection migration to the "high" capacity satellite link might be sufficient. However, I honestly don't know how this performs.

In other words, I think that the scheduler is the most crucial part for scheduling among very heterogenous links, but I also think that the scheduler has to be aware of both links at the same time. Is this sufficiently realizable with connection migration? Maybe, I don't know (yet).

**Comment by LPardue** [comment]
Devils advocate: if there were two QUIC connections, whats the downside with e.g. a client asking a server for the same thing on two different paths? The faster one will win.

What does a more integrated multipath solution solve? What is missing from QUICv1 to give that?

**Comment by joergdeutschmann-i7** [comment]
Thanks Lucas! Well, if you're asking for the same object on both paths, the terrestrial link will be fully utilized (which is bad because it is assumed to be a low data rate link) until the satellite one kicks in. In our use case both links should be utilized as little as possible. We're basically trying to combine two sub-optimal, very heterogeneous access links into one appropriate one. This is already difficult with true multipath, I'm not sure if it becomes easier without.

**Comment by larseggert** [comment]
Could we see soon a revised slide deck based on this discussion, please?

**Comment by joergdeutschmann-i7** [comment]
Updated slides are uploaded. The referenced draft-deutschmann-sat-ter-multipath-00 is now also available. Thanks!

**Comment by larseggert** [comment]
Could you answer Lucas' question, i.e., whether you require MPQIUC (i.e., `draft-deconinck`) for this use case, or whether *any* multipath QUIC design would do that conforms to certain requirements (which you should enumerate.)

**Comment by joergdeutschmann-i7** [comment]
As the slides say, we do not have tested (yet) whether connection migration might do the job, but frequently/seamlessly switching among links seems not to be the use case for connection migration. Apart from this, bandwidth aggregation is also important for low data rate terrestrial links (c.f. slides from Olivier Bonaventure).
Therefore, the only option I see is MPQUIC (draft-deconinck-quic-multipath-05) with uniflows and each uniflow having its own congestion control state.

**Comment by LPardue** [comment]
Think bigger. The other option is to design a protocol feature that support bandwidth aggregation. The MP-QUIC proposal is not the only way to achieve that.

**Comment by joergdeutschmann-i7** [comment]
Sure, I'm happy to debate other proposals. However, for example L3 bonding is also not applicable to our use case, because as soon as one packets goes over the satellite link, the whole transmission is affected by the high latency. I'm also not saying that our use case must use MPQUIC, but I think it could use MPQUIC.

**Comment by LPardue** [comment]
So is it accurate to say you'd like a QUIC-based solution for end-to-end (client - server) multipath and that you're indifferent to how that gets achieved?

**Comment by joergdeutschmann-i7** [comment]
Does that QUIC-based solution allow fine-grained scheduling decisions among heterogeneous links? Is there a candidate for such a protocol feature?

**Comment by LPardue** [comment]
That's the part of the intention of this interim, to figure out what people need or don't need. What needs to be worked on by the QUIC WG, what needs to be signalled between endpoints for interoperability, what is an implementation decision, and so on.

**Comment by joergdeutschmann-i7** [comment]
I see. So, I would say fine-grained scheduling among multiple paths and knowledge of both (or more) link characteristics simultaneously (i.e., flow and congestion control per uniflow) is needed for our use case. A requirement which was not mentioned yet is a quick connection setup (0-RTT) for both links; I've added this to the slides.

And then there is of course a more general issue, which is the performance of QUIC over geostationary satellites. Before the satellite link can be utilized in a multipath scenario, it needs to perform well in a QUIC single path scenario. I've added this to the slides, too (and honestly I think that this is even more urgent than satellite / terrestrial multipath).

**Comment by LPardue** [comment]
Thanks. I'm not hearing anything that depends on the specific proposal in draft-deconinck-quic-multipath-05. So can you please replace `MPQUIC` with `multipath QUIC` and `uniflows` with `path`.

**Comment by joergdeutschmann-i7** [comment]
Sure, sounds good to me. Thanks and let me know if I can improve anything further.

**Comment by LPardue** [comment]
Thanks @joergdeutschmann-i7 this addresses all of my comments! I'll let Lars answer if his are addressed.
